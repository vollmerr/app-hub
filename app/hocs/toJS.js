// see http://redux.js.org/docs/recipes/UsingImmutableJS.html#use-a-higher-order-component-to-convert-your-smart-components-immutablejs-props-to-your-dumb-components-javascript-props
/* eslint-disable */
import React from 'react'
import { Iterable } from 'immutable';


const toJS = (WrappedComponent) => (wrappedComponentProps) => {
  const KEY = 0;
  const VALUE = 1;

  const propsJS = Object.entries(
    wrappedComponentProps
  ).reduce((newProps, wrappedComponentProp) => {
    newProps[wrappedComponentProp[KEY]] = Iterable.isIterable(
      wrappedComponentProp[VALUE]
    )
      ? wrappedComponentProp[VALUE].toJS()
      : wrappedComponentProp[VALUE]
    return newProps;
  }, {});

  return <WrappedComponent {...propsJS} />;
};


export default toJS;
