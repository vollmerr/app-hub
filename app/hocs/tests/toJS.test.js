import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import toJS from '../toJS';


const WrappedComponent = () => <div>test component</div>;


const prop = {
  array: ['test 1', 'test 2'],
  obj: {
    a: {
      b: 'c',
    },
    n: [1, 2, 3],
  },
  bool: false,
};

const props = {
  immutable: fromJS(prop),
  plain: prop,
};


describe('toJS', () => {
  let wrapper;
  let withJS;
  beforeEach(() => {
    jest.resetAllMocks();
    withJS = toJS(WrappedComponent)(props);
    wrapper = shallow(<div>{withJS}</div>);
  });

  it('should be a HOC that returns a new component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render the Component passed', () => {
    expect(wrapper.find(WrappedComponent).length).toEqual(1);
  });

  it('should make the immutable props plain js', () => {
    expect(wrapper.find(WrappedComponent).prop('immutable')).toEqual(prop);
  });

  it('should pass the plain js props untouched', () => {
    expect(wrapper.find(WrappedComponent).prop('plain')).toEqual(prop);
  });
});

