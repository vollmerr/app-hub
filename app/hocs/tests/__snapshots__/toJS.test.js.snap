// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toJS should be a HOC that returns a new component 1`] = `
<div>
  <WrappedComponent
    immutable={
      Object {
        "array": Array [
          "test 1",
          "test 2",
        ],
        "bool": false,
        "obj": Object {
          "a": Object {
            "b": "c",
          },
          "n": Array [
            1,
            2,
            3,
          ],
        },
      }
    }
    plain={
      Object {
        "array": Array [
          "test 1",
          "test 2",
        ],
        "bool": false,
        "obj": Object {
          "a": Object {
            "b": "c",
          },
          "n": Array [
            1,
            2,
            3,
          ],
        },
      }
    }
  />
</div>
`;
